source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -0,0 +1,3510 @@
+        auto currNode = std::make_shared<UiNode>(name, id);
+        currNode->setNode(node);
+        ++id;
+        setUiNodeInfo(currNode, node->getType(), node->getCategory());
+    }
+    // creating uiNodes for the nodegraph
+    for (mx::NodeGraphPtr nodeGraph : nodeGraphs)
+    {
+        std::string name = nodeGraph->getName();
+        auto currNode = std::make_shared<UiNode>(name, id);
+        currNode->setNodeGraph(nodeGraph);
+        ++id;
+        setUiNodeInfo(currNode, "", "nodegraph");
+    }
+    for (mx::InputPtr input : inputNodes)
+    {
+        auto currNode = std::make_shared<UiNode>(input->getName(), id);
+        currNode->setInput(input);
+        ++id;
+        setUiNodeInfo(currNode, input->getType(), input->getCategory());
+    }
+    // creating edges for nodegraphs
+    for (mx::NodeGraphPtr graph : nodeGraphs)
+    {
+        for (mx::InputPtr input : graph->getActiveInputs())
+        {
+            int upNum = -1;
+            int downNum = -1;
+            mx::NodePtr connectedNode = input->getConnectedNode();
+            if (connectedNode)
+            {
+                upNum = findNode(connectedNode->getName(), "node");
+                downNum = findNode(graph->getName(), "nodegraph");
+                if (upNum > -1)
+                {
+                    Edge newEdge = Edge(_graphNodes[upNum], _graphNodes[downNum], input);
+                    _graphNodes[downNum]->edges.push_back(newEdge);
+                    _graphNodes[downNum]->setInputNodeNum(1);
+                    _graphNodes[upNum]->setOutputConnection(_graphNodes[downNum]);
+                    _currEdge.push_back(newEdge);
+                }
+            }
+        }
+    }
+    // creating edges for surface and material nodes
+    for (mx::NodePtr node : docNodes)
+    {
+        for (mx::InputPtr input : node->getActiveInputs())
+        {
+
+            mx::string nodeGraphName = input->getNodeGraphString();
+            mx::NodePtr connectedNode = input->getConnectedNode();
+            mx::OutputPtr connectedOutput = input->getConnectedOutput();
+            int upNum = -1;
+            int downNum = -1;
+            if (nodeGraphName != "")
+            {
+
+                upNum = findNode(nodeGraphName, "nodegraph");
+                downNum = findNode(node->getName(), "node");
+            }
+            else if (connectedNode)
+            {
+                upNum = findNode(connectedNode->getName(), "node");
+                downNum = findNode(node->getName(), "node");
+            }
+            else if (connectedOutput)
+            {
+                upNum = findNode(connectedOutput->getName(), "output");
+                downNum = findNode(node->getName(), "node");
+            }
+            else if (input->getInterfaceName() != "")
+            {
+                upNum = findNode(input->getInterfaceName(), "input");
+                downNum = findNode(node->getName(), "node");
+            }
+            if (upNum != -1)
```

Comment:
Right now if a graph with a channel attribute is loaded it will connect correctly but adding a channel to a port on newly added node is not currently supported.